src/inmanta/execute/proxy.py

```diff
@@ -77,27 +98,44 @@ class ProxyContext:
     :param allow_reference_values: Allow references for values accessed through this proxy. Either because they have been
         declared and validated, or because explicitly requested. Defaults to allow references iff the object has been
         validated. Kept as a separate field to allow overrides for a single proxy object while maintaining behavior for
-        nested proxies.
+        nested proxies. This value is not passed on to children.
     :param path: The path, within the plugin's argument namespace, where this object lives. Should be compatible with
         composition at the tail, e.g. `list_arg[0].attr`.
     """
 
-    path: str
-    mode: ProxyMode = ProxyMode.PLUGIN
-    validated: bool = True
-    allow_reference_values: Optional[bool] = None
+    __slots__ = ("path", "validated", "allow_reference_values")
 
-    def nested(self: Self, *, relative_path: str) -> Self:
+    def __init__(self, path: str, validated: bool = True, allow_reference_values: Optional[bool] = None) -> None:
+        self.path = path
+        self.validated = validated
+        self.allow_reference_values = allow_reference_values
+
+    def nested(self, *, relative_path: str) -> "ProxyContext":
         """
         Returns a context object for values nested one level deeper than the current context.
 
         :param relative_path: The path of new object, relative to the current context.
         """
-        return dataclasses.replace(
-            self,
+        return ProxyContext(
             path=self.path + relative_path,
-            # we're proxying elements one level deeper than the current context
-            # => reset allow_reference_values to default behavior
+            validated=self.validated,
+            allow_reference_values=None,
+        )
+
+    def allow_references(self) -> "ProxyContext":
+        return ProxyContext(
+            path=self.path,
+            validated=self.validated,
+            allow_reference_values=True,
+        )
+
+    def descend(self) -> "ProxyContext":
```

Comment:
I suspect we could get rid of this, because this behavior only applies to the base class, I would suspect it can be hardcoded, but not luck yet.

Reply:
Not sure what you mean by "hardcoded" in this context. If you mean to inline it, I think I'd prefer that, because now it's sort of arbitrary, and the name doesn't imply anything.